ldap/servers/slapd/haproxy.c

```diff
@@ -0,0 +1,392 @@
+        str_saved = slapi_ch_strdup(str);
+        copied = str_saved;
+        after_header = split_string_at_delim(str_saved, '\n');
+
+        /* Check if the header is valid */
+        if (after_header == 0) {
+            slapi_log_err(SLAPI_LOG_CONNS, "haproxy_parse_v1_hdr", "Missing protocol header terminator\n");
+            goto done;
+        }
+        /* Parse the protocol, family, addresses, and ports */
+        if (haproxy_parse_v1_protocol(get_next_token(&copied), "PROXY") < 0) {
+            slapi_log_err(SLAPI_LOG_CONNS, "haproxy_parse_v1_hdr", "Missing or bad protocol header\n");
+            goto done;
+        }
+        /* Parse the family */
+        if (haproxy_parse_v1_fam(get_next_token(&copied), &addr_family) < 0) {
+            slapi_log_err(SLAPI_LOG_CONNS, "haproxy_parse_v1_hdr", "Missing or bad protocol type\n");
+            goto done;
+        }
+        /* Parse the addresses */
+        if (haproxy_parse_v1_addr(get_next_token(&copied), &parsed_addr_from, addr_family) < 0) {
+            slapi_log_err(SLAPI_LOG_CONNS, "haproxy_parse_v1_hdr", "Missing or bad client address\n");
+            goto done;
+        }
+        if (haproxy_parse_v1_addr(get_next_token(&copied), &parsed_addr_dest, addr_family) < 0) {
+            slapi_log_err(SLAPI_LOG_CONNS, "haproxy_parse_v1_hdr", "Missing or bad server address\n");
+            goto done;
+        }
+        /* Parse the ports */
+        if (haproxy_parse_v1_port(get_next_token(&copied), &parsed_addr_from) < 0) {
+            slapi_log_err(SLAPI_LOG_CONNS, "haproxy_parse_v1_hdr", "Missing or bad client port\n");
+            goto done;
+        }
+        if (haproxy_parse_v1_port(get_next_token(&copied), &parsed_addr_dest) < 0) {
+            slapi_log_err(SLAPI_LOG_CONNS, "haproxy_parse_v1_hdr", "Missing or bad server port\n");
+            goto done;
+        }
+        rc = HAPROXY_HEADER_PARSED;
+        *proxy_connection = 1;
+        *str_len = after_header - str_saved;
+        /* Copy the parsed addresses to the output parameters */
+        memcpy(pr_netaddr_from, &parsed_addr_from, sizeof(PRNetAddr));
+        memcpy(pr_netaddr_dest, &parsed_addr_dest, sizeof(PRNetAddr));
+
+done:
+        slapi_ch_free_string(&str_saved);
+    } else {
+        rc = HAPROXY_NOT_A_HEADER;
+    }
+    return rc;
+}
+
+/**
+ * Function to receive and parse HAProxy headers, supporting both v1 and v2 of the protocol.
+ * 
+ * @param fd: The file descriptor of the socket from which to read.
+ * @param proxy_connection: A pointer to an integer to store the proxy connection status (0 or 1).
+ * @param pr_netaddr_from: A pointer to a PRNetAddr structure to store the source address info.
+ * @param pr_netaddr_dest: A pointer to a PRNetAddr structure to store the destination address info.
+ * 
+ * @return: Returns 0 on successful operation, -1 on error.
+ */
+int haproxy_receive(int fd, int *proxy_connection, PRNetAddr *pr_netaddr_from, PRNetAddr *pr_netaddr_dest)

```

Comment:
A possible case here is that you have the DS server behind haproxy, but an admin bypasses haproxy to ldap to the server. That would cause this to error because the BER doesn't decode to a haproxy header leading to a disconnect. I think that this needs to handle the haproxy header being *optional* rather than the current setup of it being required (I think ... if I'm reading it properly).

Reply:
As I read it, if the first PDU does not come through an haproxy, the first PDU match neither haproxy V1 nor V2 signature so haproxy_receive returns HAPROXY_NOT_A_HEADER then connection_read_operation will just go on processing the request happily.
So the header is already optionnal. BTW if it was not the case, most of the CI tests would have failed ! -;)
That said, there may be a loophole somewhere because some of the CI tests suspiciously fail with errors related to connection closure but I have not seen P3 or P4 disconnection code in the artifacts so it may not be related either ...

Reply:
Ahh I misread the logic in the if statement, yes, it only checks for the _ERROR case. 

CI would also be passing since haproxy parsing isn't on by default. So I think that in the haproxy = true case, there needs to be some tests for when we don't have haproxy incoming.

Reply:
A simple test for the attribute and binds is added. We can think about something more complex later. This PR already becomes too big to handle. :)

Reply:
Sounds fair but I wonder if we should not also add an environment variable that will be taken in account to set the attribute when creating the instances (i.e in _create_instances in topologies.py)
That way we should be able to easily run most of the existing tests with the haproxy enabled.

Reply:
Okay, I added the environment variable, and I've hardcoded it so our CI can run with the nsslapd-haproxy-trusted-ip attribute set (so we check for HAProxy headers). And all looks good.
The existing failures are not related, I think.

@Firstyear, please, review :)